### ISSUE TITLE
StackOverflow Error

### ISSUE THREAD

**Issue by Mathias157** [spine-tools/SpineOpt.jl#449]
Hi!

I'm currently writing my master thesis at DTU Management, Energy Systems Analysis, using a SpineOpt model. 
So first of all: Thanks for a very intuitive and flexible modelling software! I've enjoyed modelling in it, and I've been able to solve the issues that have appeared throughout my process for now.

However, I've had a stack overflow error for several days now, that I just cannot figure out how to solve. I don't have a background in software engineering so it might stem from my lack of understanding of what a "stack overflow error" actually means. 
I've tried the following:
- Scaling down the model (It's a bit large, 224 investment options [ and 672 h with hourly resolution)
- Decreasing precision of floats (max 3 decimals)
- Setting candidate_units/connections and node_state_cap previously inf, to 1000000 

Here's the exact output from console:
![image](https://user-images.githubusercontent.com/77012503/161437300-7b7faa33-9657-4615-b88d-3c7da097e2b5.png)
![image](https://user-images.githubusercontent.com/77012503/161437535-02783499-38d9-4eca-9f04-c53cfefdd745.png)
![image](https://user-images.githubusercontent.com/77012503/161437554-cee05437-65cc-41f6-8e8c-3fa389ea7095.png)
![image](https://user-images.githubusercontent.com/77012503/161437580-3b1b6575-1059-4aa5-907e-5a82782053b0.png)

Best,
Mathias

**Comment by mihlema** [comment]
Hi @Mathias157 ,
thanks a lot for reporting this issue. It's a bit tricky to diagnose, but I have an idea what might be going on here. The interesting part is probably what follows the Error message [3], `no method matching *(Nothing, Float)`. Probably what's happening here is that somewhere in the objective function we are multiplying a parameter of `Nothing` value with something else. 

To further debug this, could you maybe check which objective function terms you are including (e.g. do you have investment costs, ramping costs ... etc.?) . 

I am not sure how familiar you are with the SpineOpt code and julia, but in case you are, we could possibly find out which particular cost term is causing problems by un-/commenting the different cost terms in the `src/run_spineopt.jl` file and identify which one it is. 

If you are able to share your database (or a deceived version of it if you are using confidential data), we can also have a look at it.

**Comment by manuelma** [comment]
Hi!
Thank you for the nice words! The real issue here is a bit below the StackOverflowError. It's this: caused by MethodError, no method matching *(::Nothing, ::Float). You have a nothing in your objective function.

In these cases we attempt to print out the entire objective to the console so you can see where the nothing is and fix. Unfortunately our 'print' function is recursive and in your case, it's hiting the maximum recursion limit. We can rewrite the print function using iteration rather than recursion - I will try to do that.

**Comment by DillonJ** [comment]
Hi @Mathias157 - thanks for your feedback and for using Spine and SpineOpt - and for getting so far by yourself! 

My suspicion is that in unit_investment_costs.jl we iterate over `units_invested_available_indices()` - this function generates the indices for the investment variables and here if `candidate_units` is defined for a `unit`, then it is an investment `unit`. Same goes for connections and storages. However, we calculate the investments costs in unit_investment_costs.jl without checking if `unit_investment_cost` has been defined or not. So your problem is that you have defined `candidate_units` or `candidate_connections` but have defined no `unit_investment_cost` or `connection_investment_cost`. The work around is to just add a `unit_investment_cost`  cost of zero for the unit. But we should check for this and will update the code accordingly.

**Comment by DillonJ** [comment]
Also, on this point:

>Setting candidate_units/connections and node_state_cap previously inf, to 1000000

That is a very high value for candidate_units. If you have defined the `investment_variable_type` as an integer - it may have a big impact on performance. If this is a true reasonable estimate for the upper bound on this investment variable, then I would tend to make it a continuous variable instead.

**Comment by Mathias157** [comment]
Thanks a lot for all the swift replies - I ended up rolling back to a version that worked and started implementing the new data more systematically and now it works - it must have been a human error...

I looked though my database a number of times for lone candidate_units/connections, but haven't been able to find any. I even deleted some "alternative names" where candidates were, in fact, specified (of course, this shouldn't do anything and "luckily" it didn't).

Good suggestion to print the objective function. However, it seems run_spineopt doesn't return the model m when an error occurs? (if it successfully runs, m from m = run_spineopt is stored in the console)
